rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,962 @@
+worker, thus creating as many input pipelines as number of replicas. Each
+replica will dequeue from its own `Dataset`.
+
+#### InputContext
+
+An instance of `InputContext` is passed to the input function.
+
+```python
+class InputContext(object):
+  @property
+  def num_replicas(self) -> int:
+    """Returns the total number of compute replicas."""
+
+  @property
+  def num_replicas_in_sync(self) -> int:
+    """Returns the number of compute replicas training in sync."""
+
+  @property
+  def input_pipeline_id(self) -> int:
+    """Returns the input pipeline ID.
+
+    If `replication_mode` == `SINGLE`, this is always `0`.
+    If `replication_mode` == `PER_WORKER`, this is the worker ID.
+    If `replication_mode` == `PER_REPLICA`, this is the compute replica ID.
+    """
+
+  @property
+  def num_input_pipelines(self) -> int:
+    """Returns the number of input pipelines.
+
+    If `replication_mode` == `SINGLE`, this is always `1`.
+    If `replication_mode` == `PER_WORKER`, this is the number of workers.
+    If `replication_mode` == `PER_REPLICA`, this is `num_replicas`.
+    """
+```
+
+#### InputIterator
+
+An instance of `InputIterator` is returned by `prepare_input` and can be passed
+to `run`.
+
+```python
+class InputIterator(Generic[T]):
+  """An input iterator, intended to be passed to `Replicator.run`."""
+
+  def get_next(self) -> List[T]:
+    """Returns the next inputs for each replica."""
+
+  def reinitialize(self) -> List[tf.Operation]:
+    """Re-initializes the inputs."""
+```
+
+### Model Parallelism
+
+Model parallelism can allow the execution of models when there is insufficient
+memory on a single device (and further data parallelism is not possible).
+
+Replicator supports model partitioning, a form of model parallelism in which the
+ops within the model are placed onto multiple devices. This is done via the
+`logical_device` context.
+
+Inputs are always located on logical device 0. Ops within the step function
+default to logical device 0 (except on TPU, which defaults to automatic
+placement).
+
+```python
+with replicator.scope():
+  with replicator.logical_device_for_variables(1):
+    v = tf.get_variable(...)  # On logical device 1.
+    x = v * 3  # Ignores logical device context.
+
+def step(ctx, inputs):  # `inputs` on logical device 0.
+  a = model_part1(inputs)  # Implicitly on logical device 0.
+  with ctx.logical_device(0):
+    b = model_part2(a)  # Explicitly on logical device 0.
+  with ctx.logical_device(1):
+    return model_part3(b)  # Explicitly on logical device 1.
+```
+
+### Eager mode support
+
+Eager mode support mostly drops out of building atop of `DistributionStrategy`.
+The entire API is designed to be eager compatible. We've called out the
+difference in behavior in eager and graph mode in the specific API
+documentations. In general, in graph mode, many of the APIs return ops which
+should be later executed with `session.run`, and in eager mode, they simply run
+the ops.
+
```

Comment:
Keras supports DistStrat natively, but we have also discussed the fact that the inclusion of DistStrat in compile/fit limits our ability to handle, say, model parallelism or very large models. Is there an intersection of Keras and Replicator that allows for model parallelism/more complicated use semantics without requiring too much from the Keras user?